src/main.rs

```diff
@@ -130,16 +183,40 @@ fn optimize(matches: clap::ArgMatches) {
         }
     };
 
-    let rules = rules_from_str(split_rules);
+    // Get multi-pattern rules. learned_rules are the learned rules from TASO,
+    // pre_defined_multi are the hand-specified rules from TASO
+    let multi_patterns = if let Some(rule_file) = matches.value_of("multi_rules") {
+        let learned_rules =
+            read_to_string(rule_file).expect("Something went wrong reading the rule file");
+        let pre_defined_multi = PRE_DEFINED_MULTI.iter().map(|&x| x);
+        let multi_rules: Vec<&str> = learned_rules.split("\n").chain(pre_defined_multi).collect();
+        MultiPatterns::with_rules(multi_rules)
+    } else {
+        let multi_rules: Vec<&str> = PRE_DEFINED_MULTI.iter().map(|&x| x).collect();
+        MultiPatterns::with_rules(multi_rules)
+    };
 
     // Run saturation
-    let time_limit_sec = Duration::new(10, 0);
-    let iter_limit = 10;
+    let n_sec = matches
+        .value_of("n_sec")
+        .map_or(10, |s| s.parse::<u64>().unwrap());
+    let time_limit_sec = Duration::new(n_sec, 0);
+    let iter_limit = matches
+        .value_of("n_iter")
+        .map_or(3, |s| s.parse::<usize>().unwrap());
 
-    let runner = Runner::<Mdl, TensorAnalysis, ()>::default()
-        .with_time_limit(time_limit_sec)
-        .with_iter_limit(iter_limit)
-        .with_expr(&start);
+    let runner = if use_multi {
+        Runner::<Mdl, TensorAnalysis, ()>::default()
+            .with_time_limit(time_limit_sec)
+            .with_iter_limit(iter_limit)
+            .with_expr(&start)
+            .with_hook(move |runner| multi_patterns.run_one(runner))
```

Comment:
Can you add comment on when the multi patterns are applied?

Reply:
Done